src/merkle/mmr/full.rs

```diff
@@ -0,0 +1,289 @@
+//! A fully materialized Merkle mountain range (MMR).
+//!
+//! A MMR is a forest structure, i.e. it is an ordered set of disjoint rooted trees. The trees are
+//! ordered by size, from the most to least number of leaves. Every tree is a perfect binary tree,
+//! meaning a tree has all its leaves at the same depth, and every inner node has a branch-factor
+//! of 2 with both children set.
+//!
+//! Additionally the structure only supports adding leaves to the right-most tree, the one with the
+//! least number of leaves. The structure preserves the invariant that each tree has different
+//! depths, i.e. as part of adding adding a new element to the forest the trees with same depth are
+//! merged, creating a new tree with depth d+1, this process is continued until the property is
+//! restabilished.
+use super::bit::TrueBitPositionIterator;
+use super::{super::Vec, Accumulator, Opening, Partial, Rpo256, Word};
+
+// MMR
+// ===============================================================================================
+
+/// A fully materialized Merkle Mountain Range, with every tree in the forest and all their
+/// elements.
+///
+/// Since this is a full representation of the MMR, elements are never removed and the MMR will
+/// grow roughly `O(2n)` in number of leaf elements.
+pub struct FullMMR {
+    /// Refer to the `forest` method documentation for details of the semantics of this value.
+    pub(super) forest: usize,
+
+    /// Contains every element of the forest.
+    ///
+    /// The trees are in postorder sequential representation. This representation allows for all
+    /// the elements of every tree in the forest to be stored in the same sequential buffer. It
+    /// also means new elements can be added to the forest, and merging of trees is very cheap with
+    /// no need to copy elements.
+    pub(super) nodes: Vec<Word>,
+}
+
+impl Default for FullMMR {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl FullMMR {
+    // CONSTRUCTORS
+    // ============================================================================================
+
+    /// Constructor for an empty `FullMMR`.
+    pub fn new() -> FullMMR {
+        FullMMR {
+            forest: 0,
+            nodes: Vec::new(),
+        }
+    }
+
+    // ACCESSORS
+    // ============================================================================================
+
+    /// Returns the MMR forest representation.
+    ///
+    /// The forest value has the following interpretations:
+    /// - its value is the number of elements in the forest
+    /// - bit count corresponds to the number of trees in the forest
+    /// - each true bit position determines the depth of a tree in the forest
+    pub const fn forest(&self) -> usize {
+        self.forest
```

Comment:
MMR is a forest data structure, the `forest` attribute is a bitmask representing the trees in the forest. For example `0b1010` means there are two trees in the MMR, the left most tree has `8` elements and the smaller `2` elements.

Because the trees in the MMR are perfect, i.e. number of leaves is a power of two, the bitmaks happens to coincide with the leaf count of the MMR. So the two are one and the same, the name change is mostly to make the algorithms easier to read/write. For an Accumulator instance we only really care the number of leaves in the MMR. For the MMR we care about the tree position and its individual size.

I did take quite a bit of care while naming the variables, because I had a few bugs caused by type confusion.

- `forest`: number of trees and their size in the MMR
- `tree`: a single tree with a power-of-two number of leafs
- `offset`: an index to the underlying storage vector, includes leaves and internal nodes.
- `position`: a leaf position, does not include internal nodes